src/api/app/controllers/webui/project_controller.rb

```diff
@@ -387,68 +387,20 @@ def move_path
   end
 
   def monitor
-    @legend = Buildresult::STATUS_DESCRIPTION
-
-    @name_filter = params[:pkgname]
-    @lastbuild_switch = params[:lastbuild]
-    if params[:defaults]
-      defaults = (begin
-                    Integer(params[:defaults])
-                  rescue ArgumentError
-                    1
-                  end) > 0
-    else
-      defaults = true
-    end
-    params['expansionerror'] = 1 if params['unresolvable']
-    monitor_set_filter(defaults)
-
-    find_opt = { project: @project, view: 'status', code: @status_filter,
-                 arch: @arch_filter, repository: @repo_filter }
-    find_opt[:lastbuild] = 1 if @lastbuild_switch.present?
-
-    @buildresult = Buildresult.find_hashed(find_opt)
-    if @buildresult.empty?
-      flash[:warning] = "No build results for project '#{@project}'"
-      redirect_to action: :show, project: params[:project]
-      return
-    end
-
-    unless @buildresult.key?('result')
+    unless (buildresult = monitor_buildresult)
       @buildresult_unavailable = true
       return
     end
 
-    fill_status_cache
+    monitor_parse_buildresult(buildresult)
 
-    @localpackages = {}
-    @project.packages.each do |package|
-      @localpackages[package.name] = 1
+    # extract repos
+    repohash = {}
+    @statushash.each do |repo, arch_hash|
+      repohash[repo] = arch_hash.keys.sort!
```

Comment:
This is less efficient than before... you need to iterate once to get the keys and another onw to sort... and you are using this only twice... `sort!` in the first iteration it is more efficient

Reply:
What first iteration? the arch_hash is a hash and the archs are its keys.